portfolio/src/main/java/com/google/sps/servlets/DeleteDataServlet.java

```diff
@@ -46,11 +42,15 @@ public void init() {
   
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    Query commentKeysQuery = new Query("Comment").setKeysOnly();
-    PreparedQuery commentKeysResults =  datastore.prepare(commentKeysQuery);
-    for(Entity commentKey : commentKeysResults.asIterable()){
-      datastore.delete(commentKey.getKey());
+    // Verify the deleter is the author of the comment being deleted. If not, early return.
+    UserService userService = UserServiceFactory.getUserService();
+    if (!userService.isUserLoggedIn()
+        || !request
+            .getParameter("authorEmail")
```

Comment:
This is not safe because you are trusting the request about which person authored the comment. You should instead send only the ID in the request, then read the comment from Datastore and check the author on the entity in Datastore.